proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,1294 @@
+   ``type`` herald at the term level).
+
+5. When ``ViewPatterns`` are enabled, interpret ``f (a -> b) = ...``
+   as a view pattern, otherwise as ``f ((->) a b) = ...``.
+
+6. ``case ... of x -> y -> z`` is an error. We require parentheses to
+   disambiguate:
+
+   * ``case ... of (x -> y) -> z``
+   * ``case ... of x -> (y -> z)``
+
+Name resolution
+~~~~~~~~~~~~~~~
+
+7. During name resolution,
+
+   * Identifiers bound in term syntax populate the term namespace;
+     identifiers bound in type syntax populate the type namespace.
+
+     This is already the case, but now we generalize this rule to cover
+     types-in-terms, which are considered term syntax.
+
+   * When looking up an identifier ``v`` or ``V`` in type syntax, look it up
+     in the type namespace first; if it is not found there, look it up in the
+     term namespace.
+
+     This is already the case for uppercase identifiers if ``DataKinds`` is
+     enabled, but now we extend this rule to lowercase identifiers if
+     ``RequiredTypeArguments`` is enabled.
+
+   * When looking up an identifier ``v`` or ``V`` in term syntax, look it up
+     in the term namespace first; if it is not found there, look it up in the
+     type namespace.
+
+     This is a new rule, but notice how it mirrors the one for type syntax.
+
+Implicit quantification
+~~~~~~~~~~~~~~~~~~~~~~~
+
+8. Implicit quantification is an existing feature that allows the programmer to
+   omit a ``forall``::
+
+     g ::           a -> a    -- implicit
+     g :: forall a. a -> a    -- explicit
+
+   This sort of quantification only happens if the variable is not already in
+   scope::
+
+     {-# LANGUAGE ScopedTypeVariables #-}
+
+     f :: forall a. a -> a
+     f = ...
+       where
+         g :: a -> a         -- No implicit quantification!
+
+   In other words, we quantify only over *free* variables.
+
+   With the proposed changes to name resolution, variables that were previously
+   free are not free anymore::
+
+       a = 42
+       f :: a -> a           -- No implicit quantification!
+
+   This is a breaking change, and that is why the fallback to the term
+   namespace in type syntax is guarded behind ``RequiredTypeArguments``.
+
+   Without ``RequiredTypeArguments``, implicit quantification is not affected.
+
+   In order to facilitate writing code that is forward-compatible with
+   ``RequiredTypeArguments``, introduce a new warning to ``-Wcompat``: ``-Wterm-variable-capture``.
+   This warning will notify users when implicit quantification occurs that
+   would stop working under ``RequiredTypeArguments``.
+
+Type checking
+~~~~~~~~~~~~~
+
+9. Introduce an additional typing rule that transforms term arguments into type
+   arguments::
+
+     rho = t2t(e)
+     G |- (forall a -> sigma); type rho, pis  ~>  Theta; phis; rho_r
+     ----------------------------------------------------------------  T2T
+     G |- (forall a -> sigma);        e, pis  ~>  Theta; phis; rho_r
```

Comment:
It was one rule initially, but Simon found it more clear to separate concerns.

Reply:
If it comes to this choice, it's probably better to make me grumpy rather than @simonpj , indeed. I still prefer the one rule version though :slightly_smiling_face: .

Reply:
Arnaud is right.   (Maybe I misunderstood or was just plain wrong before.)

We should *replace* the rule ITVDQ from Section 4 with the new rule
```
     rho = t2t(e)
     G |- sigma[a := rho]; pis  ~>  Theta; phis; rho_r
     ----------------------------------------------------------------  ITVDQ
     G |- (forall a -> sigma);        e, pis  ~>  Theta; phis; rho_r```
```
Then in 7.6 we need a prominent bullet to say that `t2t( type t ) = t`.  And a comment on the new rule to point out that it is identical to the old one when `e` is `type t`.

Apart from economy of rules, we need that `t2t` bullet anyway, because you can always write
```
f (Int -> type (Bool -> Bool))
```
where the outer `Int ->` is term syntax, but the inner `Bool -> Bool` is type syntax.